database-composite-types/README.md

```diff
@@ -0,0 +1,172 @@
+
+```python
+class PipelineModel(Base):
+    __tablename__ = "pipelines"
+    id = Column(Integer, primary_key=True)
+    datetime = Column(DateTime, default=datetime.utcnow)
+
+    job_trigger_id = Column(Integer, ForeignKey("job_triggers.id"))
+    job_trigger = relationship("JobTriggerModel", back_populates="runs")
+
+    srpm_build_id = Column(Integer, ForeignKey("srpm_builds.id"), index=True)
+    srpm_build = relationship("SRPMBuildModel", back_populates="runs")
+    copr_build_group_id = Column(
+        Integer, ForeignKey("copr_build_groups.id"), index=True
+    )
+    copr_build_group = relationship("CoprBuildGroupModel", back_populates="runs")
+    test_run_group_id = Column(
+        Integer, ForeignKey("tft_test_run_groups.id"), index=True
+    )
+    test_run_group = relationship("TFTTestRunGroupModel", back_populates="runs")
+
+```
+
+- has the foreign keys to other tables to reference steps of the pipeline (groups), the group then references particular
+  targets, therefore when getting the whole pipeline, join on multiple tables has to happen
+
+## 1.option
+
+How could the model look like when using composite type on target level:
+
+- storing the data for the particular target directly in pipeline
+- this would be a step back since we already did the grouping refactoring
+
+### Issues
+
+#### 1. querying concrete steps of the pipeline
+
+Examples of when this happens:
+
+- when updating Copr build status in DB, we get the corresponding builds via the build ID from Copr (index)
+- when updating TF run status in DB, we get the corresponding test run via the TF pipeline ID (index)
+- babysit tasks - they get all the pending Copr builds / Test runs
+- when triggering `/packit test` - we get the latest Copr build model with corresponding commit SHA
+
+#### 2. data duplication
+
+- updating SRPM build would require getting all pipelines and updating the data everywhere
+
+##### Average number of Copr builds using the same SRPM build
+
+```
+packit=# SELECT AVG(copr_build_target_count) FROM (
+    SELECT COUNT(DISTINCT copr_build_targets.id) AS copr_build_target_count
+    FROM pipelines
+    JOIN copr_build_targets ON pipelines.copr_build_group_id = copr_build_targets.copr_build_group_id
+    GROUP BY pipelines.srpm_build_id
+) AS copr_targets_count;
+        avg
+--------------------
+ 6.4208079446515274
+(1 row)
+```
+
+- similar situation for Copr builds <> TF runs, since the relationship can be 1:n
+
+##### Average number of tests using the same Copr build
+
+```
+packit=# SELECT AVG(test_run_count) AS average_test_runs FROM (
+    SELECT copr_id, COUNT(DISTINCT tft_id) AS test_run_count
+    FROM tf_copr_build_association_table
+    GROUP BY copr_id
+) AS tf_counts;
+ average_test_runs
+--------------------
+ 1.2618628164375168
+(1 row)
+```
+
+## 2.option
+
+How could the model look like when using composite type on group level:
+
+- "groups" stored together in one pipeline
+- mapping build -> TF needed
+
+### Issues
```

Comment:
I see only issues, are there any pros as well? :wink:

Reply:
sorry, I thought I included it in the beginning (added now); as the issue suggests, the idea was to make the database design pipeline-centric and remove the need to do the joins on multiple tables